dvc/remote/base.py

```diff
@@ -680,20 +680,49 @@ def safe_remove(self, path_info, force=False):
         self.remove(path_info)
 
     def _checkout_file(
-        self, path_info, checksum, force, progress_callback=None
+        self,
+        path_info,
+        checksum,
+        force,
+        progress_callback=None,
+        save_link=True,
     ):
-        cache_info = self.checksum_to_path_info(checksum)
-        if self.exists(path_info):
-            msg = "data '{}' exists. Removing before checkout."
-            logger.warning(msg.format(str(path_info)))
-            self.safe_remove(path_info, force=force)
+        # NOTE: In case if path_info is already cached and path_info's
+        # link type matches cache link type, we would like to avoid
+        # relinking.
+        if self.changed(
+            path_info, {self.PARAM_CHECKSUM: checksum}
+        ) or not self._is_same_link_as_cache(path_info):
+            self._remove_and_relink(path_info, checksum, force, save_link)
+        else:
+            self._restore_protection_state(path_info)
 
-        self.link(cache_info, path_info)
-        self.state.save_link(path_info)
-        self.state.save(path_info, checksum)
         if progress_callback:
             progress_callback(str(path_info))
 
+    def _is_same_link_as_cache(self, path_info):
+        return True
+
+    def _remove_and_relink(self, path_info, checksum, force, save_link):
+        self.safe_remove(path_info, force=force)
+
+        cache_info = self.checksum_to_path_info(checksum)
+        self.link(cache_info, path_info)
+
+        if save_link:
+            self.state.save_link(path_info)
+
+        self.state.save(path_info, checksum)
+
+    def _restore_protection_state(self, path_info):
+        # NOTE: performing (un)protection costs us +/- the same as checking if
+        # path_info is protected. Instead of implementing logic,
+        # just (un)protect according self.protected.
+        if self.protected:
+            self.protect(path_info)
+        else:
+            self.unprotect(path_info, allow_copy=False)
```

Comment:
It would be nice to clarify why we don't need to copy in this case.

Reply:
done